### PR TITLE
Amcl laser received callback

### LINKED ISSUE #169: Refactor laserReceived method of AMCL
[WIP] This method is very long and it can possibly be refactored into several descriptive functions to improve code readability.

### PR DESCRIPTION
**Pull request by mhpanah**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | fixes #169  |
| Primary OS tested on | Ubuntu 18.04 |
| Primary platform tested on | HP Desktop |

--- 

## Description of contribution in a few bullet points

--- 
**Will merge after internal release.**

This PR re-opens PR #310 which addresses issue #169.

- AMCL's `laserReceived` CB is long and hard to follow.  In this PR, I've created few helper functions for `laserReceived` to improve the readability. 

## Future work that may be required in bullet points

---

### CODE REVIEW

File: nav2_amcl/src/amcl_node.cpp [hunk 1]
```diff
@@ -624,134 +594,39 @@ AmclNode::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
...
-    if (update) {
-      for (unsigned int i = 0; i < lasers_update_.size(); i++) {
-        lasers_update_[i] = true;
-      }
-    }
-  }
-
   bool force_publication = false;
   if (!pf_init_) {
-    // Pose at last filter update
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp]: I'd leave comments here where possible. AMCL is a pretty considerably complex thing to try to learn reading code without some hints.

**Comment by mkhansen-intel** [comment]
Looks like code coverage has gone down 8%, can we add unit tests? I'm not even sure I understand that data so @crdelsey should first verify the coverage report.

**Comment by crdelsey** [comment]
@mhpanah Does this introduce any functional changes besides change a DEBUG print to an ERROR print?

**Comment by crdelsey** [comment]
> Looks like code coverage has gone down 8%

I suspect that our code coverage statistics are unstable because our tests are unstable. Sometimes our system tests fail and we retry them; sometimes they pass on the first try. I suspect that when we fail and retry we hit additional code that we missed when we succeeded on our first try. This can cause a PR to show unrealistic changes in code coverage results. If the last build on master had a fail and retry result and the PR passed on the first try, code coverage looks like it went down. If it's the other way around, code coverage looks like it went up.

**Comment by crdelsey** [comment]
@mhpanah The Travis build failed on nav2_util lifecycle unit tests. I've never seen that failure before. I've restarted build to see if it will pass on the retry.
